File: samples/android/app/src/androidTest/java/org/mozilla/samples/gleancore/pings/DeletionRequestPingTest.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.samples.gleancore.pings

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.action.ViewActions.closeSoftKeyboard
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.ext.junit.rules.ActivityScenarioRule
import org.mozilla.samples.gleancore.R

import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mozilla.samples.gleancore.MainActivity
import androidx.test.uiautomator.UiDevice
import mozilla.telemetry.glean.testing.GleanTestLocalServer
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertEquals

@RunWith(AndroidJUnit4::class)
class DeletionRequestPingTest {
    private val server = createMockWebServer()

    @get:Rule
    val activityRule: ActivityScenarioRule<MainActivity> = ActivityScenarioRule(MainActivity::class.java)

    @get:Rule
    val gleanRule = GleanTestLocalServer(context, server.port)

    private val context: Context
        get() = ApplicationProvider.getApplicationContext()

    @Test
    fun validateDeletionRequestPing() {
        // Wait for the app to be idle/ready.
        InstrumentationRegistry.getInstrumentation().waitForIdleSync()
        val device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
        device.waitForIdle()

        // Wait for any ping to make sure there are no pending requests before going forward
        waitForPingContent("", null, server)

        // Disable upload by toggling the switch
        onView(withId(R.id.uploadSwitch))
            .perform(closeSoftKeyboard())
            .perform(click())

        // We must get the deletion request on the first attempt,
        // no other ping should be sent after disabling upload
        val deletionPing = waitForPingContent("deletion-request", null, server, 1)!!

        // Validate the received data.

        var clientInfo = deletionPing.getJSONObject("client_info")
        val clientId = clientInfo.getString("client_id")
        assertNotEquals(clientId, "c0ffeec0-ffee-c0ff-eec0-ffeec0ffeec0")

        val metrics = deletionPing.getJSONObject("metrics")
        val uuids = metrics.getJSONObject("uuid")
        val legacyId = uuids.getString("legacy_ids.client_id")
        assertEquals("01234567-89ab-cdef-0123-456789abcdef", legacyId)

        // Try re-enabling and waiting for next baseline ping
        onView(withId(R.id.uploadSwitch)).perform(click())

        // Move it to background.
        device.pressHome()

        // Validate the received data.
        val baselinePing = waitForPingContent("baseline", null, server)!!

        clientInfo = baselinePing.getJSONObject("client_info")

        val newClientId = clientInfo.getString("client_id")
        assertNotEquals(newClientId, clientId)
        assertNotEquals(newClientId, "c0ffeec0-ffee-c0ff-eec0-ffeec0ffeec0")
    }
}

File: samples/ios/app/glean-sample-appUITests/DeletionRequestPingTest.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import Glean
import Swifter
import XCTest

// swiftlint:disable force_cast
// REASON: Used in below test cases to cause errors if data is missing
class DeletionRequestPingTest: XCTestCase {
    var app: XCUIApplication!
    var expectation: XCTestExpectation?
    var lastPingJson: [String: Any]?

    override func setUp() {
        // In UI tests it is usually best to stop immediately when a failure occurs.
        continueAfterFailure = false

        // UI tests must launch the application that they test.
        // Doing this in setup will make sure it happens for each test method.
        app = XCUIApplication()
    }

    override func tearDown() {
        self.lastPingJson = nil
        self.expectation = nil
    }

    func setupServer(expectPingType: String, port: Int = 0) -> HttpServer {
        return mockServer(expectPingType: expectPingType, port: UInt16(port)) { json in
            self.lastPingJson = json
            // Fulfill test's expectation once we parsed the incoming data.
            self.expectation?.fulfill()
        }
    }

    func testDeletionRequestPing() {
        var server = setupServer(expectPingType: "deletion-request")
        expectation = expectation(description: "Completed upload")
        let port = try! server.port()

        app.launchArguments = ["USE_MOCK_SERVER", "\(port)"]
        app.launch()

        app.switches.firstMatch.tap()

        waitForExpectations(timeout: 5.0) { error in
            XCTAssertNil(error, "Test timed out waiting for upload: \(error!)")
        }

        var clientInfo = lastPingJson!["client_info"] as! [String: Any]
        let clientId = clientInfo["client_id"] as! String
        XCTAssertNotEqual(clientId, "c0ffeec0-ffee-c0ff-eec0-ffeec0ffeec0")

        // Test deletion-request legacy id payload
        let metrics = lastPingJson!["metrics"] as! [String: Any]
        let uuids = metrics["uuid"] as! [String: Any]
        let legacyId = uuids["legacy_ids.client_id"] as! String
        XCTAssertEqual("01234567-89ab-cdef-0123-456789abcdef", legacyId)

        server.stop()

        // Try re-enabling and waiting for next baseline ping
        server = setupServer(expectPingType: "baseline", port: port)
        expectation = expectation(description: "Completed upload")

        app.switches.firstMatch.tap()

        // Trigger baseline ping by putting app into the background
        XCUIDevice.shared.press(XCUIDevice.Button.home)

        waitForExpectations(timeout: 5.0) { error in
            XCTAssertNil(error, "Test timed out waiting for upload: \(error!)")
        }

        clientInfo = lastPingJson!["client_info"] as! [String: Any]
        let newClientId = clientInfo["client_id"] as! String
        XCTAssertNotEqual(newClientId, clientId)
        XCTAssertNotEqual(newClientId, "c0ffeec0-ffee-c0ff-eec0-ffeec0ffeec0")

        server.stop()
    }
}
